### PR TITLE
Allow event monitoring to ignore failing event accessors

### LINKED ISSUE #1948: Exceptions in event accessors fail tests when using .Monitor()
### Description

When an event handler accessor has an explicit implementation, which throws an exception on the setter, `.Monitor()` is failing the test completely.

**What is my motivation?**

I need to test the class behavior without changing the other base event handler accessors behavior. It could cause unwanted side effects. But I do not need to test this event. I just wanted to test PropertyChanged events being raised. Therefore, I would like to have the benefit of having the full class available to write expressions in the assertion.

**Suggestion:**

Make some more configuration options for the Monitor. Like ignoring failing events, or just subscribing to a subset of interfaces with those events, or putting some events on ignore, without changing the generic Type, which is used to assert / monitor.

**Benefit:**

Using the great syntax you introduced in monitor on brown field projects or if you are not able to change the behavior of some event accessors. Therefore, write better and cleaner test code from the beginning.

### Complete minimal example reproducing the issue

```csharp
using System.ComponentModel;
using Xunit;

namespace FluentAssertions.Tests;

class FailingEventHandlerSubscription : INotifyPropertyChanged
{
    private event PropertyChangedEventHandler? InnerPropertyChanged;
    public event PropertyChangedEventHandler? PropertyChanged
    {
        add { throw new System.Exception("This is sadly happening due to anonymous method"); }
        remove { InnerPropertyChanged -= value; }
    }

    private object prop;

    public object Property
    {
        get => prop;
        set
        {
            prop = value;
            this.InnerPropertyChanged?.Invoke(this, new PropertyChangedEventArgs(nameof(Property)));
        }
    }
}

public class UnitTest2
{
    [Fact]
    public void TestEventHandler()
    {
        // Arrange
        var cut = new FailingEventHandlerSubscription();

       // will fail, because the exception is thrown, when the EventHandler class is subscribing the event.
        using var monitored = cut.Monitor();

        // Act
        cut.Property = new object();

        // Assert
        // If I just use the INotifyPropertyChanged interface, I cannot write this awesome expression.
        monitored.Should().RaisePropertyChangeFor(m => m.Property);
    }
}
```

### Expected behavior:

FluentAssertions is able to have options in the monitor to ignore failing event handlers.

### Actual behavior:

It throws an exception when .Monitor() is called and I can do nothing about it, event though I do not want to test this specific event.

### Versions

FluentAssertions Version 6.7.0

### Additional Information

I already implemented my custom implementation of the monitor and would offer to contribute my suggestion above to the FluentAssertions project. I just wanted to talk to you first, if you would approve of such an enhancement of the monitor or if you disagree on this feature for some reason.

So let me know if you agree or disagree.

cheers!

### PR DESCRIPTION
**Pull request by apazureck**
Added options to event monitor and added option to suprress exceptions on event accessors.

Closes #1948

## IMPORTANT 

* [x] The code complies with the [Coding Guidelines for C#](https://www.csharpcodingguidelines.com/).
* [x] The changes are covered by unit tests which follow the Arrange-Act-Assert syntax and the naming conventions such as is used [in these tests](../tree/develop/Tests/FluentAssertions.Equivalency.Specs/MemberMatchingSpecs.cs#L51-L430).
* [x] If the PR adds a feature or fixes a bug, please update [the release notes](../tree/develop/docs/_pages/releases.md) with a functional description that explains what the change means to consumers of this library, which are published on the [website](https://fluentassertions.com/releases).
* [x] If the PR changes the public API the changes needs to be included by running [AcceptApiChanges.ps1](../tree/develop/AcceptApiChanges.ps1) or [AcceptApiChanges.sh](../tree/develop/AcceptApiChanges.sh).
* [x] If the PR affects [the documentation](../tree/develop/docs/_pages), please include your changes in this pull request so the documentation will appear on the [website](https://www.fluentassertions.com/introduction).

### CODE REVIEW

File: Tests/FluentAssertions.Specs/Events/EventMonitorSpecs.cs [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+            //Assert
+            classToMonitor.Invoking(c =>
+            {
+                using var monitor = c.Monitor<IRemoveFailingEvent>(opt => opt.IgnoreEventAccessorExceptions());
+            }).Should().NotThrow();
+        }
+    }
+
+    public class SettingTimestampProvider
```
Comment (dennisdoomen) [inline Tests/FluentAssertions.Specs/Events/EventMonitorSpecs.cs:83]: 🔧 These tests are not needed. The only reason why we allow passing a custom date/time provider is so that we can decouple the tests from date/time issues.

**Review by dennisdoomen — CHANGES_REQUESTED** [review]

File: Src/FluentAssertions/Events/EventMonitorOptions.cs [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+{
+    /// <summary>
+    /// Will ignore the events, if they throw an exception on any custom event accessor implementation. default: false.
+    /// </summary>
+    internal bool ShouldIgnoreEventAccessorExceptions { get; private set; }
+
+    /// <summary>
+    /// Func used to generate the timestamp.
+    /// </summary>
+    internal Func<DateTime> TimestampProvider { get; private set; } = () => DateTime.Now;
```
[the remark this answers is not in the document]
Comment (apazureck) [reply Src/FluentAssertions/Events/EventMonitorOptions.cs:19]:
```suggestion
    internal Func<DateTime> TimestampProvider { get; private set; } = () => DateTime.UtcNow;
```

File: Tests/FluentAssertions.Specs/FluentAssertions.Specs.csproj [hunk 3]
```diff
@@ -53,6 +53,7 @@
 
   <ItemGroup>
     <PackageReference Include="Chill" Version="4.1.0" />
+    <PackageReference Include="Moq" Version="4.18.1" />
```
[the remark this answers is not in the document]
Comment (apazureck) [reply Tests/FluentAssertions.Specs/FluentAssertions.Specs.csproj:56]:
```suggestion
```

File: docs/_pages/eventmonitoring.md [hunk 4]
```diff
@@ -90,6 +96,28 @@ metadata.Should().BeEquivalentTo(new[]
 });
 ```
 
+## Configuring the event monitor
+
+It is also possible to configure the event monitor. To do so, use the overload:
+
+```csharp
+var subject = new EditCustomerViewModel(options => ...);
+using var monitoredSubject = subject.Monitor();
```
[the remark this answers is not in the document]
Comment (apazureck) [reply docs/_pages/eventmonitoring.md:105]: No, it is complete nonsense, I put it in EditCustomerViewModel....

File: docs/_pages/releases.md [hunk 5]
```diff
@@ -10,6 +10,8 @@ sidebar:
 ## Unreleased
 
 ### What's new
+* Added options to `.Monitor(...)` extension - [#1954](https://github.com/fluentassertions/fluentassertions/pull/1954)
```
[the remark this answers is not in the document]
Comment (apazureck) [reply docs/_pages/releases.md:13]: I saw those as two separate topics:

1. Make options for the EventMonitor / make it configurable
2. Ignore Exceptions on event accessors

[the remark this answers is not in the document]
Comment (dennisdoomen) [reply docs/_pages/releases.md:13]: The first is the way you made the second possible. So I wouldn't mention it.

**Comment by apazureck** [comment]
Hi,

Quick question: I did changes on my fork and pushed it. Sadly github does not seem to detect them in this PR. Do you know how I can update the PR?

**Comment by dennisdoomen** [comment]
Are these not your changes?

image

**Comment by apazureck** [comment]
Ah ok... odd. This did not show up, that's why I was a bit confused (hadn't used github in some time...)

File: docs/_pages/eventmonitoring.md [hunk 6]
```diff
@@ -90,6 +96,28 @@ metadata.Should().BeEquivalentTo(new[]
...
+```
+
+### Ignore broken event Accessors
+
+If some event accessors, you are not interested in, should throw exceptions, you can simply ignore all exceptions thrown by those event acessors:
+
+```csharp
+var subject = new EditCustomerViewModel(options => options.IgnoreEventAccessorExceptions());
```
[the remark this answers is not in the document]
Comment (apazureck) [reply docs/_pages/eventmonitoring.md]: Changed that.

File: docs/_pages/releases.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (apazureck) [reply docs/_pages/releases.md:13]: Fixed

File: Src/FluentAssertions/Events/EventMonitorOptions.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+    /// </summary>
+    /// <returns>The options instance for method stacking.</returns>
+    public EventMonitorOptions IgnoreEventAccessorExceptions()
+    {
+        ShouldIgnoreEventAccessorExceptions = true;
+        return this;
+    }
+
+    /// <summary>
+    /// Sets the timestamp provider. default it is <see cref="DateTime.Now"/>.
```
[the remark this answers is not in the document]
Comment (apazureck) [reply Src/FluentAssertions/Events/EventMonitorOptions.cs:32]: Fixed.

File: Src/FluentAssertions/Events/EventMonitorOptions.cs [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    /// </summary>
+    /// <param name="timestampProvider">The timestamp provider.</param>
+    /// <returns>The options instance for method stacking.</returns>
+    public EventMonitorOptions ConfigureTimestampProvider(Func<DateTime> timestampProvider)
```
[the remark this answers is not in the document]
Comment (apazureck) [reply Src/FluentAssertions/Events/EventMonitorOptions.cs:36]: Fixed.

File: Tests/FluentAssertions.Specs/Events/EventMonitorSpecs.cs [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+using System.Linq;
+using System.Reflection;
+using Xunit;
+
+#pragma warning disable CS0067
+
+namespace FluentAssertions.Specs.Events;
+
+[Collection("EventMonitoring")]
+public class EventMonitorSpecs
```
[the remark this answers is not in the document]
Comment (apazureck) [reply Tests/FluentAssertions.Specs/Events/EventMonitorSpecs.cs:16]: Here I also thought of SRP. This is not about asserting any events, but about subscribing them. But I'll just move it to the other class.

File: Tests/FluentAssertions.Specs/Events/EventMonitorSpecs.cs (hunk 1 above)
Comment (apazureck) [reply Tests/FluentAssertions.Specs/Events/EventMonitorSpecs.cs:83]: done

File: Tests/FluentAssertions.Specs/Events/EventMonitorSpecs.cs [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+                using var monitor = c.Monitor<IRemoveFailingEvent>();
+            }).Should().Throw<TargetInvocationException>();
+        }
+    }
+
+    public class IgnoreMisbehavingEventAccessors
+    {
+        [Fact]
+        public void
+            When_IgnoreEventAccessorExceptions_is_set_monitoring_class_with_broken_event_add_accessor_should_not_fail_test()
```
[the remark this answers is not in the document]
Comment (apazureck) [reply Tests/FluentAssertions.Specs/Events/EventMonitorSpecs.cs:53]: fixed

File: Tests/FluentAssertions.Specs/Events/EventMonitorSpecs.cs [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+
+namespace FluentAssertions.Specs.Events;
+
+[Collection("EventMonitoring")]
+public class EventMonitorSpecs
+{
+    public class MonitorDefaultBehavior
+    {
+        [Fact]
+        public void When_default_configuration_is_active_broken_event_add_accessors_should_fail_the_test()
```
[the remark this answers is not in the document]
Comment (apazureck) [reply Tests/FluentAssertions.Specs/Events/EventMonitorSpecs.cs:21]: fixed on all tests

File: Src/FluentAssertions/Events/EventMonitor.cs [hunk 12]
```diff
@@ -117,20 +120,67 @@ public void Dispose()
...
+            }
+            catch
+            {
+                // ignore
+            }
+        }
+        else
+        {
+            recorder.Dispose();
+        }
```
Comment (jnyrup) [inline Src/FluentAssertions/Events/EventMonitor.cs:145]:
```cs
try
{
    recorder.Dispose();
}
catch when (options.ShouldIgnoreEventAccessorExceptions)
{
    // ignore
}
```

File: Src/FluentAssertions/Events/EventMonitor.cs [hunk 13]
```diff
@@ -117,20 +120,67 @@ public void Dispose()
...
+    private void AttachEventHandler(EventInfo eventInfo, EventRecorder recorder)
+    {
+        if (options.ShouldIgnoreEventAccessorExceptions)
+        {
+            AttachEventHandlerOrRemoveFromRecorderMapIfAttachmentFailed(eventInfo, recorder);
+        }
+        else
+        {
+            recorder.Attach(subject, eventInfo);
+        }
```
Comment (jnyrup) [inline Src/FluentAssertions/Events/EventMonitor.cs:169]:
```cs
try
{
    recorder.Attach(subject, eventInfo);
}
catch when (options.ShouldIgnoreEventAccessorExceptions)
{
    if (!recorderMap.TryRemove(eventInfo.Name, out _))
    {
        throw new InvalidOperationException(
            $"Could not remove event {eventInfo.Name} with broken event accessor from event recording.");
    }
}
```

File: Src/FluentAssertions/AssertionExtensions.cs [hunk 14]
```diff
@@ -916,7 +919,25 @@ public static TaskCompletionSourceAssertions<T> Should<T>(this TaskCompletionSou
...
+    /// </summary>
+    /// <param name="eventSource">The object for which to monitor the events.</param>
+    /// <param name="configureOptions">
+    /// Options to configure the EventMonitor.
+    /// </param>
+    /// <exception cref="ArgumentNullException">Thrown if <paramref name="eventSource"/> is Null.</exception>
+    public static IMonitor<T> Monitor<T>(this T eventSource, Action<EventMonitorOptions> configureOptions)
+    {
+        var options = new EventMonitorOptions();
+        configureOptions?.Invoke(options);
```
Comment (jnyrup) [inline Src/FluentAssertions/AssertionExtensions.cs:939]: In most (all?) other assertions we call `Guard.ThrowIfArgumentIsNull(configureOptions, nameof(configureOptions));` instead of allowing a null value.

File: Src/FluentAssertions/Events/EventMonitor.cs [hunk 15]
```diff
@@ -19,18 +19,21 @@ internal class EventMonitor<T> : IMonitor<T>
     private readonly ConcurrentDictionary<string, EventRecorder> recorderMap =
         new ConcurrentDictionary<string, EventRecorder>();
 
-    public EventMonitor(object eventSource, Func<DateTime> utcNow)
+    public EventMonitor(object eventSource, EventMonitorOptions options)
     {
         Guard.ThrowIfArgumentIsNull(eventSource, nameof(eventSource), "Cannot monitor the events of a <null> object.");
 
+        this.options = options ?? new EventMonitorOptions();
```
Comment (jnyrup) [inline Src/FluentAssertions/Events/EventMonitor.cs:26]: Is there a real use-case for accepting a null `options`?

File: Tests/FluentAssertions.Specs/Events/EventAssertionSpecs.cs [hunk 16]
```diff
@@ -718,7 +713,8 @@ private object CreateProxyObject()
                 AssemblyBuilderAccess.Run);
             ModuleBuilder moduleBuilder = assemblyBuilder.DefineDynamicModule(assemblyName.Name, false);
             string typeName = baseType.Name + "_GeneratedForTest";
-            TypeBuilder typeBuilder = moduleBuilder.DefineType(typeName, TypeAttributes.Public, baseType, new[] { interfaceType });
+            TypeBuilder typeBuilder =
+ moduleBuilder.DefineType(typeName, TypeAttributes.Public, baseType, new[] { interfaceType });
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Events/EventAssertionSpecs.cs:717]:
```suggestion
            TypeBuilder typeBuilder =
                 moduleBuilder.DefineType(typeName, TypeAttributes.Public, baseType, new[] { interfaceType });
```

**Review by dennisdoomen — APPROVED** [review]

File: Src/FluentAssertions/Events/EventMonitor.cs (hunk 12 above)
Comment (apazureck) [reply Src/FluentAssertions/Events/EventMonitor.cs:145]: Wow, great improvement! Thanks! I'd never thought doing it with catch when.

File: Src/FluentAssertions/AssertionExtensions.cs (hunk 14 above)
Comment (apazureck) [reply Src/FluentAssertions/AssertionExtensions.cs:939]: fixed

File: Src/FluentAssertions/Events/EventMonitor.cs (hunk 15 above)
Comment (apazureck) [reply Src/FluentAssertions/Events/EventMonitor.cs:26]: fixed

File: Src/FluentAssertions/Events/EventMonitor.cs [hunk 17]
```diff
@@ -19,18 +19,22 @@ internal class EventMonitor<T> : IMonitor<T>
     private readonly ConcurrentDictionary<string, EventRecorder> recorderMap =
         new ConcurrentDictionary<string, EventRecorder>();
 
-    public EventMonitor(object eventSource, Func<DateTime> utcNow)
+    public EventMonitor(object eventSource, EventMonitorOptions options)
     {
         Guard.ThrowIfArgumentIsNull(eventSource, nameof(eventSource), "Cannot monitor the events of a <null> object.");
+        Guard.ThrowIfArgumentIsNull(eventSource, nameof(eventSource), "Event monitor needs configuration.");
```
Comment (IT-VBFK) [inline Src/FluentAssertions/Events/EventMonitor.cs:25]: Was this intentional? (2nd `Guard` is the same)

File: Src/FluentAssertions/Events/EventMonitorOptions.cs [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    /// </summary>
+    /// <returns>The options instance for method stacking.</returns>
+    public EventMonitorOptions IgnoreEventAccessorExceptions()
+    {
+        ShouldIgnoreEventAccessorExceptions = true;
+        return this;
+    }
+
+    /// <summary>
+    /// Sets the timestamp provider. default it is <see cref="DateTime.Now"/>.
```
Comment (jnyrup) [inline Src/FluentAssertions/Events/EventMonitorOptions.cs:32]:
```suggestion
    /// Sets the timestamp provider. By default it is <see cref="DateTime.UtcNow"/>.
```

File: Src/FluentAssertions/Events/EventMonitor.cs (hunk 17 above)
Comment (jnyrup) [reply Src/FluentAssertions/Events/EventMonitor.cs:25]:
```suggestion
        Guard.ThrowIfArgumentIsNull(options, nameof(options), "Event monitor needs configuration.");
```

**Review by jnyrup — CHANGES_REQUESTED** [review]

File: Src/FluentAssertions/Events/EventMonitor.cs [hunk 19]
```diff
@@ -117,20 +121,48 @@ public void Dispose()
...
+        {
+            recorder.Attach(subject, eventInfo);
+        }
+        catch when (options.ShouldIgnoreEventAccessorExceptions)
+        {
+            if (!recorderMap.TryRemove(eventInfo.Name, out _))
+            {
+                throw new InvalidOperationException(
+                    $"Could not remove event {eventInfo.Name} with broken event accessor from event recording.");
             }
```
Comment (jnyrup) [inline Src/FluentAssertions/Events/EventMonitor.cs:166]: For my understanding, why do we need this?
I tried to remove it and no tests failed.

I get that the `Try` part is hard to exercise because of concurrency, but I guess(?) that not removing from `recorderMap` at should be testable?

Comment (jnyrup) [reply Src/FluentAssertions/Events/EventMonitor.cs:166]: @apazureck Have you had a chance to look into this?

Comment (apazureck) [reply Src/FluentAssertions/Events/EventMonitor.cs:166]: Hi, sorry for the long delay, I somehow overlooked this message before I was on holiday the last weeks.

You are right, I did not test this explicitly, as it is internal and I cannot provoke the error without manipulating the internal state of the `EventMonitor` via reflection.

I thought it would be good to have an exception, if some code will change and this will be removed in the meantime. I cannot think of any valid reason to do this, but the person, who did the change, should be alerted, that this is not an expected behavior. I thought this is better than silently swallow the inconsistent behavior.

But I can write another test this week manipulating the eventmonitor, or I can remove this exception. Tell me what you think is the better approach.

Comment (jnyrup) [reply Src/FluentAssertions/Events/EventMonitor.cs:166]: Great to have you back 😄

Not sure if we're in sync here.

1)
Why are we removing the event from `recorderMap` in the first place?

I removed these lines and no tests failed which indicates that there's a scenario we're not covering.
```cs
if (!recorderMap.TryRemove(eventInfo.Name, out _))
{
    throw new InvalidOperationException(
        $"Could not remove event {eventInfo.Name} with broken event accessor from event recording.");
}
```

This can be tested with something like this.
```cs
var subject = new TestEventBrokenEventHandlerRaising();
using var monitor = subject.Monitor<IAddFailingEvent>(opt => opt.IgnoreEventAccessorExceptions());
monitor.GetRecordingFor(nameof(IAddFailingEvent.AddFailingEvent))
```

If we remove the event from `recorderMap`, then `GetRecordingFor` throws an `InvalidOperationException` that we're not monitoring `IAddFailingEvent`, otherwise we get a empty recording.
What do we want here?

I haven't used these monitoring APIs in my day job, so I'm not _that_ familiar with all the ins and outs and what is expected here.

2)
This question is only relevant if the outcome of 1) is to keep `TryRemove`.

As I understand `TryRemove` (just had to look it up) a false means that:
* The value was already removed when calling `TryRemove`
* Another concurrent call removed the value, 
  * I.e. for N concurrent calls to `TryRemove` on an existing key, 1 returns true and N-1 return false.

In `AttachEventHandler` we don't throw an exception if `TryAdd` returns false, so I don't think we should throw here.

Comment (jnyrup) [reply Src/FluentAssertions/Events/EventMonitor.cs:166]: @dennisdoomen can you help here, since you know more about Events than me

File: Src/FluentAssertions/Events/EventMonitor.cs [hunk 20]
```diff
@@ -117,20 +121,48 @@ public void Dispose()
...
+
+    private void AttachEventHandler(EventInfo eventInfo, EventRecorder recorder)
+    {
+        try
+        {
+            recorder.Attach(subject, eventInfo);
+        }
+        catch when (options.ShouldIgnoreEventAccessorExceptions)
+        {
+            if (!recorderMap.TryRemove(eventInfo.Name, out _))
```
Comment (dennisdoomen) [inline Src/FluentAssertions/Events/EventMonitor.cs:162]: 🔧 Instead of removing the event from the map at this point (which is what I think caused the confusion), have this method return a `bool`. And then in caller, do something like 

```csharp
            if (recorderMap.TryAdd(eventInfo.Name, recorder))
            {
                if (!AttachEventHandler(eventInfo, recorder))
                {
                    recorderMap.TryRemove(eventInfo.Name, recorder);
                }
            }
```

I don't think we should be throwing exceptions. Somebody _could_ have disposed the monitor.

**Comment by dennisdoomen** [comment]
Any progress on this @apazureck ?

**Comment by apazureck** [comment]
Hi,

I am so sorry, I totally procrastinated on that one. I will update the PR until end of February and then you can hopefully pull it in.

**Comment by dennisdoomen** [comment]
@apazureck we'd love to pull this in, but only if you finish it.

File: Src/FluentAssertions/Events/EventMonitor.cs (hunk 20 above)
Comment (apazureck) [reply Src/FluentAssertions/Events/EventMonitor.cs:163]: Hi, sorry for the long delay. I checked out the code and when you call the AttachEventHandler with the EventRecorder argument and this calles tryremove anyway, then the second try remove does not do anything, as far as I understand. Because you are calling the tryremove two times in a row, it returns the true and then you call the try again.

The use case here is: Ignore the error when a custom event accessor throws any exception on attaching. When it is not on the recorder map and thus not removed from it, then it is fine.

TLDR: I just skipped throwing the exception when the EventRecorder was not in the recorderMap dictionary, or it was removed twice, because I don't think it matters anyway for the tests.

**Comment by apazureck** [comment]
Hi,

Sorry for the long absence. I threw out the exception when the event recorder was not on the dictionary and I added an option to still get events, even if the event add accessor did throw an exception. It is possible to set the event to the internal event list and throw an exception afterwards. So I hope this will cover most cases of this exotic topic.

**Review by dennisdoomen — APPROVED** [review]

File: docs/_pages/eventmonitoring.md [hunk 21]
```diff
@@ -90,6 +96,28 @@ metadata.Should().BeEquivalentTo(new[]
...
+It is also possible to configure the event monitor. To do so, use the overload:
+
+```csharp
+var subject = new EditCustomerViewModel();
+using var monitoredSubject = subject.Monitor(options => ...);
+```
+
+### Ignore broken event Accessors
+
+If some event accessors, you are not interested in, should throw exceptions, you can simply ignore all exceptions thrown by those event acessors:
```
Comment (IT-VBFK) [inline docs/_pages/eventmonitoring.md:110]:
```suggestion
If some event accessors, you are not interested in, should throw exceptions, you can simply ignore all exceptions thrown by those event accessors:
```

File: docs/_pages/eventmonitoring.md [hunk 22]
```diff
@@ -90,6 +96,28 @@ metadata.Should().BeEquivalentTo(new[]
...
+
+```csharp
+var subject = new EditCustomerViewModel();
+using var monitoredSubject = subject.Monitor(options => options.IgnoreEventAccessorExceptions());
+```
+
+The event monitor will now ignore all events that throw exceptions when subscribing (`add`) or unsubscribing (`remove`) them.
+
+> Note, that if there is an exception in the `add` event acessor, the event cannot be asserted, as it cannot be subscribed successfully.
```
Comment (IT-VBFK) [inline docs/_pages/eventmonitoring.md:119]:
```suggestion
> Note, that if there is an exception in the `add` event accessor, the event cannot be asserted, as it cannot be subscribed successfully.
```

File: docs/_pages/eventmonitoring.md (hunk 21 above)
Comment (IT-VBFK) [reply docs/_pages/eventmonitoring.md:110]: See NUKE spellcheck

File: docs/_pages/eventmonitoring.md (hunk 22 above)
Comment (IT-VBFK) [reply docs/_pages/eventmonitoring.md:119]: See NUKE spellcheck

**Comment by IT-VBFK** [comment]
And API-checks..

File: Tests/FluentAssertions.Specs/Events/EventAssertionSpecs.cs [hunk 23]
```diff
@@ -971,6 +972,151 @@ public void One_matching_argument_type_with_two_or_more_parameters_matching_a_mi
...
+            monitor.MonitoredEvents.Should().HaveCount(1);
+        }
+
+        [Fact]
+        public void Ignoring_broken_event_accessor_should_also_not_record_events()
+        {
+            // Arrange
+            var classToMonitor = new TestEventBrokenEventHandlerRaising();
+
+            using var monitor = classToMonitor.Monitor<IAddFailingEvent>(opt => opt.IgnoreEventAccessorExceptions(false));
```
Comment (jnyrup) [inline Tests/FluentAssertions.Specs/Events/EventAssertionSpecs.cs:1053]: `IgnoreEventAccessorExceptions` now takes a Boolean parameter which we'd like to avoid in the APIs.
See https://csharpcodingguidelines.com/maintainability-guidelines/#AV1564 for an explanation.

I assume you combined them, as `recordEventsWithBrokenAccessor` only has an effect when ignoring event accessor exceptions?

**Comment by dennisdoomen** [comment]
@apazureck any change you will finish this PR?

**Comment by IT-VBFK** [comment]
Would this go into 6.12 if someone finishes it?

**Comment by apazureck** [comment]
Hi,

I will fix the stuff until Friday. Thought it was OK and already finished.

File: Tests/FluentAssertions.Specs/Events/EventAssertionSpecs.cs (hunk 23 above)
Comment (IT-VBFK) [reply Tests/FluentAssertions.Specs/Events/EventAssertionSpecs.cs:1053]: @apazureck I think this one is still open :)

**Comment by IT-VBFK** [comment]
Can you also take care of the remaining qodana issues?

**Comment by apazureck** [comment]
> Can you also take care of the remaining qodana issues?

Hi,

Yes, I am on it. But the job took so long to run I went to bed yesterday.

File: docs/_pages/eventmonitoring.md (hunk 21 above)
Comment (IT-VBFK) [reply docs/_pages/eventmonitoring.md:110]: ping

File: Tests/FluentAssertions.Specs/Events/EventAssertionSpecs.cs (hunk 23 above)
Comment (apazureck) [reply Tests/FluentAssertions.Specs/Events/EventAssertionSpecs.cs]: Added another method to set recroding to true, even if event setter fails.

File: Src/FluentAssertions/AssertionExtensions.cs [hunk 24]
```diff
@@ -917,7 +917,31 @@ public static TaskCompletionSourceAssertions<T> Should<T>(this TaskCompletionSou
...
-        return new EventMonitor<T>(eventSource, utcNow ?? (() => DateTime.UtcNow));
+        var options = new EventMonitorOptions();
+
+        if (utcNow is not null)
+        {
+            options.ConfigureTimestampProvider(utcNow);
+        }
+
+        return new EventMonitor<T>(eventSource, options);
+    }
```
Comment (jnyrup) [inline Src/FluentAssertions/AssertionExtensions.cs:928]: Now that we're open to change the existing API, we could remove `Func<DateTime> utcNow = null`, such that this becomes.
```cs
public static IMonitor<T> Monitor<T>(this T eventSource) =>
    new EventMonitor<T>(eventSource, new EventMonitorOptions());
```

For testing we still have the internal `ConfigureTimestampProvider`.

@dennisdoomen What do you think?

Comment (dennisdoomen) [reply Src/FluentAssertions/AssertionExtensions.cs:928]: > For testing we still have the internal `ConfigureTimestampProvider`.

What's that?

Comment (jnyrup) [reply Src/FluentAssertions/AssertionExtensions.cs:928]: `EventMonitorOptions.ConfigureTimestampProvider`.

In #625 the `Func<DateTime> utcNow = null` was added, but I suspect it wasn't meant for public use, but merely a necessity for testing.

Comment (dennisdoomen) [reply Src/FluentAssertions/AssertionExtensions.cs:928]: Ah yes, you're right. That would no longer be necessary.

Comment (apazureck) [reply Src/FluentAssertions/AssertionExtensions.cs:928]: Do you have a workflow for obsolete methods? Should it be tagged obsolete first with the remart it will be removed in the next version? Then people have some time, if they are using it for some reason.

Comment (dennisdoomen) [reply Src/FluentAssertions/AssertionExtensions.cs:928]: Yes, that's how we normally do it. But last week, the `develop` branch is used for the next major version, so we're fine to introduce well-argumented breaking changes in that branch.

Comment (apazureck) [reply Src/FluentAssertions/AssertionExtensions.cs:928]: I removed the utcnow. I handed down the options to the eventrecorder, as otherwise the late changed reference on the timestamp provider would not be used. The events are subscribed when the constructor is called. Please check if it is conform to your other code and let me know, if it you like another approach.

Comment (ITaluone) [reply Src/FluentAssertions/AssertionExtensions.cs:928]: > In https://github.com/fluentassertions/fluentassertions/pull/625 the Func<DateTime> utcNow = null was added, but I suspect it wasn't meant for public use, but merely a necessity for testing.

We should mention this in the docs as well :)

**Comment by dennisdoomen** [comment]
You'll need to rebase your changes on `develop` to resolve the conflicts.

**Comment by IT-VBFK** [comment]
Ping

**Comment by apazureck** [comment]
Thanks for the reminder, I thought it was already finished. I can tend to this on the upcoming weekend and hope it will pass then.

**Comment by ITaluone** [comment]
Pong

**Comment by ITaluone** [comment]
@apazureck Do you have a plan on finishing this?
